### PR TITLE
Add cache-snapshot support

### PR DESCRIPTION
**Pull request by tylerwilliams**
Snapshots are read and written from the cache as directories containing all of the necessary files (kernel, initrd, containerfs, workspacefs, etc). This does not yet attach a different workspace to a snapshot-resumed VM.

### CODE REVIEW

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 1]
```diff
@@ -83,7 +88,7 @@ const (
 
 	// Workspace slack space is how much extra space will be allocated in the
 	// workspace disk image beyond the size of the existing files.
-	workspaceSlackBytes = 1e9
+	workspaceSlackBytes = 100 * 1e6 // 100MB
```
Comment (siggisim) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:91]: is this effectively a limit on the size of the outputs or am i misunderstanding what this is?

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go [hunk 2]
```diff
@@ -341,36 +283,168 @@ func NewContainer(ctx context.Context, opts ContainerOpts) (*firecrackerContaine
...
+		c.vmIdx = opts.ForceVMIdx
+	}
 	return c, nil
 }
 
+func (c *FirecrackerContainer) SaveSnapshot(ctx context.Context, instanceName string) (string, error) {
+	if err := c.machine.PauseVM(ctx); err != nil {
+		return "", err
+	}
+	memSnapshotPath := filepath.Join(c.getChroot(), fullDiskSnapshotName)
```
Comment (siggisim) [inline enterprise/server/remote_execution/containers/firecracker/firecracker.go:322]: i think mem and disk naming is swapped here

**Review by siggisim — APPROVED** [review]
looks good, just a couple of ubernits

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 2 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go]: Done.

File: enterprise/server/remote_execution/containers/firecracker/firecracker.go (hunk 1 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/containers/firecracker/firecracker.go:91]: Until fuse lands, yup :(

**Pull request merged**